File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/time/TimeFormatter.java
package com.mapbox.services.android.navigation.v5.utils.time;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Typeface;
import android.text.SpannableStringBuilder;
import android.text.style.RelativeSizeSpan;
import android.text.style.StyleSpan;

import com.mapbox.services.android.navigation.R;
import com.mapbox.services.android.navigation.v5.navigation.NavigationTimeFormat;
import com.mapbox.services.android.navigation.v5.utils.span.SpanItem;
import com.mapbox.services.android.navigation.v5.utils.span.SpanUtils;
import com.mapbox.services.android.navigation.v5.utils.span.TextSpanItem;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class TimeFormatter {

  private static final String TIME_STRING_FORMAT = " %s ";

  public static String formatTime(Calendar time, double routeDuration, @NavigationTimeFormat.Type int type,
                                  boolean isDeviceTwentyFourHourFormat) {
    time.add(Calendar.SECOND, (int) routeDuration);
    TimeFormattingChain chain = new TimeFormattingChain();
    return chain.setup(isDeviceTwentyFourHourFormat).obtainTimeFormatted(type, time);
  }

  public static SpannableStringBuilder formatTimeRemaining(Context context, double routeDuration) {
    long seconds = (long) routeDuration;

    if (seconds < 0) {
      throw new IllegalArgumentException("Duration must be greater than zero.");
    }

    long days = TimeUnit.SECONDS.toDays(seconds);
    seconds -= TimeUnit.DAYS.toSeconds(days);
    long hours = TimeUnit.SECONDS.toHours(seconds);
    seconds -= TimeUnit.HOURS.toSeconds(hours);
    long minutes = TimeUnit.SECONDS.toMinutes(seconds);
    seconds -= TimeUnit.MINUTES.toSeconds(minutes);

    if (seconds >= 30) {
      minutes = minutes + 1;
    }

    List<SpanItem> textSpanItems = new ArrayList<>();
    Resources resources = context.getResources();
    formatDays(resources, days, textSpanItems);
    formatHours(context, hours, textSpanItems);
    formatMinutes(context, minutes, textSpanItems);
    formatNoData(context, days, hours, minutes, textSpanItems);
    return SpanUtils.combineSpans(textSpanItems);
  }

  private static void formatDays(Resources resources, long days, List<SpanItem> textSpanItems) {
    if (days != 0) {
      String dayQuantityString = resources.getQuantityString(R.plurals.numberOfDays, (int) days);
      String dayString = String.format(TIME_STRING_FORMAT, dayQuantityString);
      textSpanItems.add(new TextSpanItem(new StyleSpan(Typeface.BOLD), String.valueOf(days)));
      textSpanItems.add(new TextSpanItem(new RelativeSizeSpan(1f), dayString));
    }
  }

  private static void formatHours(Context context, long hours, List<SpanItem> textSpanItems) {
    if (hours != 0) {
      String hourString = String.format(TIME_STRING_FORMAT, context.getString(R.string.hr));
      textSpanItems.add(new TextSpanItem(new StyleSpan(Typeface.BOLD), String.valueOf(hours)));
      textSpanItems.add(new TextSpanItem(new RelativeSizeSpan(1f), hourString));
    }
  }

  private static void formatMinutes(Context context, long minutes, List<SpanItem> textSpanItems) {
    if (minutes != 0) {
      String minuteString = String.format(TIME_STRING_FORMAT, context.getString(R.string.min));
      textSpanItems.add(new TextSpanItem(new StyleSpan(Typeface.BOLD), String.valueOf(minutes)));
      textSpanItems.add(new TextSpanItem(new RelativeSizeSpan(1f), minuteString));
    }
  }

  private static void formatNoData(Context context, long days, long hours, long minutes,
                                   List<SpanItem> textSpanItems) {
    if (days == 0 && hours == 0 && minutes == 0) {
      String minuteString = String.format(TIME_STRING_FORMAT, context.getString(R.string.min));
      textSpanItems.add(new TextSpanItem(new StyleSpan(Typeface.BOLD), String.valueOf(1)));
      textSpanItems.add(new TextSpanItem(new RelativeSizeSpan(1f), minuteString));
    }
  }
}

File: libtrip-notification/src/main/java/com/mapbox/navigation/trip/notification/utils/time/TimeFormatter.kt
package com.mapbox.navigation.trip.notification.utils.time

import android.content.Context
import android.content.res.Resources
import android.graphics.Typeface
import android.text.SpannableStringBuilder
import android.text.style.RelativeSizeSpan
import android.text.style.StyleSpan
import com.mapbox.navigation.base.typedef.TimeFormatType
import com.mapbox.navigation.trip.notification.R
import com.mapbox.navigation.utils.extensions.combineSpan
import com.mapbox.navigation.utils.span.SpanItem
import com.mapbox.navigation.utils.span.TextSpanItem
import java.util.ArrayList
import java.util.Calendar
import java.util.Locale
import java.util.concurrent.TimeUnit

/**
 * Default Time Formatter
 */
object TimeFormatter {

    private const val TIME_STRING_FORMAT = " %s "

    /**
     * Format time
     *
     * @param time Calendar
     * @param routeDuration duration in seconds
     * @param type [TimeFormatType]
     * @param isDeviceTwentyFourHourFormat *true* if 24-hour format, *false* otherwise
     * @return String
     */
    @JvmStatic
    fun formatTime(
        time: Calendar,
        routeDuration: Double,
        @TimeFormatType type: Int,
        isDeviceTwentyFourHourFormat: Boolean
    ): String {
        time.add(Calendar.SECOND, routeDuration.toInt())
        val chain = TimeFormattingChain()
        return chain.setup(isDeviceTwentyFourHourFormat).obtainTimeFormatted(type, time)
    }

    /**
     * Format remaining time
     *
     * @param context Context
     * @param routeDuration route duration in seconds
     * @return SpannableStringBuilder
     */
    @JvmStatic
    fun formatTimeRemaining(context: Context, routeDuration: Double, locale: Locale?): SpannableStringBuilder {
        var seconds = routeDuration.toLong()

        if (seconds < 0) {
            seconds = 0L
        }

        val days = TimeUnit.SECONDS.toDays(seconds)
        seconds -= TimeUnit.DAYS.toSeconds(days)
        val hours = TimeUnit.SECONDS.toHours(seconds)
        seconds -= TimeUnit.HOURS.toSeconds(hours)
        val minutes =
            TimeUnit.SECONDS.toMinutes(seconds + TimeUnit.MINUTES.toSeconds(1) / 2) // round it to next minute if seconds is more or equal than 30

        val textSpanItems = ArrayList<SpanItem>()
        val resources = context.resourcesWithLocale(locale)
        formatDays(resources, days, textSpanItems)
        formatHours(resources, hours, textSpanItems)
        formatMinutes(resources, minutes, textSpanItems)
        formatNoData(resources, days, hours, minutes, textSpanItems)
        return textSpanItems.combineSpan()
    }

    private fun formatDays(resources: Resources, days: Long, textSpanItems: MutableList<SpanItem>) {
        if (days != 0L) {
            val dayQuantityString =
                resources.getQuantityString(R.plurals.numberOfDays, days.toInt())
            val dayString = String.format(TIME_STRING_FORMAT, dayQuantityString)
            textSpanItems.add(TextSpanItem(StyleSpan(Typeface.BOLD), days.toString()))
            textSpanItems.add(TextSpanItem(RelativeSizeSpan(1f), dayString))
        }
    }

    private fun formatHours(resources: Resources, hours: Long, textSpanItems: MutableList<SpanItem>) {
        if (hours != 0L) {
            val hourString = String.format(TIME_STRING_FORMAT, resources.getString(R.string.hr))
            textSpanItems.add(TextSpanItem(StyleSpan(Typeface.BOLD), hours.toString()))
            textSpanItems.add(TextSpanItem(RelativeSizeSpan(1f), hourString))
        }
    }

    private fun formatMinutes(
        resources: Resources,
        minutes: Long,
        textSpanItems: MutableList<SpanItem>
    ) {
        if (minutes != 0L) {
            val minuteString = String.format(TIME_STRING_FORMAT, resources.getString(R.string.min))
            textSpanItems.add(TextSpanItem(StyleSpan(Typeface.BOLD), minutes.toString()))
            textSpanItems.add(TextSpanItem(RelativeSizeSpan(1f), minuteString))
        }
    }

    private fun formatNoData(
        resources: Resources,
        days: Long,
        hours: Long,
        minutes: Long,
        textSpanItems: MutableList<SpanItem>
    ) {
        if (days == 0L && hours == 0L && minutes == 0L) {
            val minuteString = String.format(TIME_STRING_FORMAT, resources.getString(R.string.min))
            textSpanItems.add(TextSpanItem(StyleSpan(Typeface.BOLD), 1.toString()))
            textSpanItems.add(TextSpanItem(RelativeSizeSpan(1f), minuteString))
        }
    }

    private fun Context.resourcesWithLocale(locale: Locale?): Resources {
        val config = this.resources.configuration.also {
            it.setLocale(locale)
        }
        return this.createConfigurationContext(config).resources
    }
}
